Add support for arm cca

MatiasVara opened the pull request:
This PR aims at adding support to build realm guests. First commit adds support for `create_guest_memfd()` and `set_user_memory_region2()`. To do this, the `memory_init()` is modified by adding a boolean parameter. This is required when building a confidential guest for arm cca and probably also required by other cases.
The second commit imports the `virtee/cca` crate and adds the steps to build a cca guest. The following items should be completed before merge the PR:
- [x] use populate() only for kernel and initialize() for non-kernel area
- [x] handle the case in which the guest switch from shared  -> private
- [x] correctly calculation of max_ipa

This has been testing using the v7 series for Linux as a guest and v8 series for KVM on FVP model.

Feedback is welcome.

MatiasVara committed:
Support ARM CCA feature

Enable to build confidential guests using ARM CCA (Confidential
Computing Architecture). This work relies on v7 series for Linux and v8
series for KVM. This has been tested only on the corresponding FVP model
simulator. For testing, you require specific kvm-ioctls and kvm-bindings
crates.

Signed-off-by: Matias Ezequiel Vara Larsen <mvaralar@redhat.com>

jakecorrenti commented:
@MatiasVara is this ready for review, or should I wait?

tylerfanelli commented:
@jakecorrenti I'm addressing the KVM guest_memfd changes (which @MatiasVara also added in this PR) to my latest SEV-SNP patches. This will probably require a rebase after that. After that, it will likely be ready for a review.

MatiasVara commented:
> @jakecorrenti I'm addressing the KVM guest_memfd changes (which @MatiasVara also added in this PR) to my latest SEV-SNP patches. This will probably require a rebase after that. Once that happens, it will likely be ready for a review.

I think the PR is ready to review. The current issue is that it does not work for the latest series for KVM (v6). I am investigating the issue.

MatiasVara commented:
I just updated the PR with the fix for v6 series. The fix is only to set the `KVM_ARM_VCPU_REC` feature during the initialization of VCPUs otherwise `vcpu_finalize` would fail.

MatiasVara commented:
@jakecorrenti @tylerfanelli @slp Feel free to give a quick look. I am pretty sure there many things to fix.

jakecorrenti commented on src/vmm/src/builder.rs:
```diff
@@ -72,6 +78,11 @@ use vm_memory::Bytes;
 
+use vm_memory::GuestMemoryRegion;
+
+use crossbeam_channel::Sender;
```
We already import this for macos

jakecorrenti commented on src/vmm/src/builder.rs:
```diff
@@ -72,6 +78,11 @@ use vm_memory::Bytes;
+use vm_memory::GuestMemoryRegion;
+
+use crossbeam_channel::Sender;
+use kvm_bindings::KVM_ARM_VCPU_REC;
```
```suggestion
#[cfg(feature = "cca")]
use kvm_bindings::KVM_ARM_VCPU_REC;
```

jakecorrenti commented on src/vmm/src/builder.rs:
```diff
@@ -674,7 +731,7 @@ pub fn build_microvm(
         attach_unixsock_vsock_device(&mut vmm, vsock, event_manager, intc.clone())?;
         #[cfg(not(feature = "net"))]
-        vmm.kernel_cmdline.insert_str("tsi_hijack")?;
+        //vmm.kernel_cmdline.insert_str("tsi_hijack")?;
         #[cfg(feature = "net")]
```
Should this be CCA only? There's another one a few lines below as well

MatiasVara replied:
I can't remember why I commented this line.

jakecorrenti commented on src/vmm/src/linux/vstate.rs:
```diff
@@ -45,7 +50,12 @@ use kvm_bindings::{
     KVM_MAX_CPUID_ENTRIES, KVM_PIT_SPEAKER_DUMMY,
 };
-use kvm_bindings::{kvm_userspace_memory_region, KVM_API_VERSION};
+use kvm_bindings::{
+    kvm_create_guest_memfd, kvm_memory_attributes, kvm_userspace_memory_region,
+    kvm_userspace_memory_region2, KVM_API_VERSION, KVM_MEMORY_ATTRIBUTE_PRIVATE,
+    KVM_MEMORY_EXIT_FLAG_PRIVATE, KVM_MEM_GUEST_MEMFD, KVM_VM_TYPE_ARM_IPA_SIZE_MASK,
+    KVM_VM_TYPE_ARM_REALM,
+};
```
Arm-specific bindings shouldn't be imported for all flavors

MatiasVara replied:
Should I import them only when `#[cfg(feature = "tee")]`?

jakecorrenti commented on src/vmm/src/linux/vstate.rs:
```diff
@@ -1228,12 +1347,36 @@ impl Vcpu {
                     error!("Received KVM_EXIT_FAIL_ENTRY signal: reason={reason}, vcpu={vcpu}");
                     Err(Error::VcpuUnhandledKvmExit)
                 }
+                // TODO: to remove this
+                Unsupported(39) => Ok(VcpuEmulation::Handled),
```
leftover TODO

MatiasVara replied:
After checking this, I figured out this should remain. I'll explain in the code.

jakecorrenti commented on src/vmm/src/linux/vstate.rs:
```diff
@@ -927,7 +1035,12 @@ impl Vcpu {
+        id: u8,
+        vm_fd: &VmFd,
+        exit_evt: EventFd,
+        sender_io: Sender<MemProperties>,
```
should be CCA only

tylerfanelli replied:
The memory fault exception isn't TEE-specific (although it could be used that way), so I don't think I agree that this should be CCA only.

jakecorrenti replied:
You're right. I did this part of the review before you and I had talked about the order of patches.

MatiasVara replied:
I think we can enable the extra this extra thread only when the `tee` feature is enabled.

jakecorrenti commented on src/vmm/src/builder.rs:
```diff
@@ -1103,17 +1231,19 @@ fn create_vcpus_aarch64(
     guest_mem: &GuestMemoryMmap,
     entry_addr: GuestAddress,
     exit_evt: &EventFd,
+    sender_io: Sender<MemProperties>,
```
should be CCA only

jakecorrenti reviewed (changes requested):
I didn't actually run it, just looked at the code. Overall looks good. Few comments, though.

Thanks for the work!

jakecorrenti commented on src/arch/Cargo.toml:
```diff
@@ -18,8 +20,8 @@ smbios = { path = "../smbios" }
 [target.'cfg(target_os = "linux")'.dependencies]
-kvm-bindings = { version = ">=0.8", features = ["fam-wrappers"] }
-kvm-ioctls = ">=0.17"
+kvm-bindings = { version = ">=0.8", features = ["fam-wrappers"] , git = "https://github.com/virtee/kvm-bindings", branch = "add_bindings_for_realms" }
+kvm-ioctls = { version = ">=0.17", git = "https://github.com/virtee/kvm-ioctls", branch = "cca" }
```
Exactly. I'm going to have to make some changes for my TDX work, so I think we'll need to merge your CCA work and my TDX work into the main branch and then use that here.

MatiasVara commented on src/libkrun/src/lib.rs:
```diff
@@ -1225,9 +1235,12 @@ pub extern "C" fn krun_start_enter(ctx_id: u32) -> i32 {
     #[cfg(target_os = "macos")]
     let (sender, receiver) = unbounded();
 
+    let (io_sender, receiver) = unbounded();
+
     let _vmm = match vmm::builder::build_microvm(
         &ctx_cfg.vmr,
         &mut event_manager,
+        io_sender,
         ctx_cfg.shutdown_efd,
```
I am not sure about the design either.  I do not like that all flavors spawn an extra thread so I think I will change that. Why couldn't you have a function signature as you proposed?

MatiasVara commented on src/libkrun/src/lib.rs:
```diff
@@ -1242,6 +1255,61 @@ pub extern "C" fn krun_start_enter(ctx_id: u32) -> i32 {
+                    for (_index, region) in guest_mem.iter().enumerate() {
+                        if (region.start_addr().raw_value() + region.size() as u64) > m.addr {
+                            let offset = m.addr - region.start_addr().raw_value();
+                            let host_startaddr = m.addr + offset;
```
Yes, it is based on `ram_block_discard_range()`. You are right, it should use actual host address. Thanks.

MatiasVara commented on src/libkrun/src/lib.rs:
```diff
@@ -1242,6 +1255,61 @@ pub extern "C" fn krun_start_enter(ctx_id: u32) -> i32 {
+                            let offset = m.addr - region.start_addr().raw_value();
+                            let host_startaddr = m.addr + offset;
+                            unsafe {
+                                let _ret = madvise(
```
Right. How do I handle the error in this case?

jakecorrenti replied:
I don't know how to handle this off the top of my head, but at the very least I think we should probably check if the ret value is < 0 and then log the error

MatiasVara replied:
Yes, I will check for that condition.

MatiasVara commented on src/devices/Cargo.toml:
```diff
@@ -5,7 +5,9 @@ authors = ["The Chromium OS Authors"]
 edition = "2021"
 
 [features]
+default = ["cca"]
 tee = []
+cca = []
```
Should I not leave `tee =[]` here?

jakecorrenti replied:
You should. I think that was a suggestion error. I was trying to suggest just getting rid of `default = ...`

MatiasVara commented on src/arch/Cargo.toml:
```diff
@@ -5,6 +5,8 @@ authors = ["The Chromium OS Authors"]
 edition = "2021"
 
 [features]
+default = ["cca"]
+cca = []
```
If I understand correctly, when I compile with cca flags:
```bash
cargo build --release --features cca
```
The `tee` feature would be enabled. However, most of the code for `tee` is not required and I ended up doing this everywhere:
```diff
-    #[cfg(feature = "tee")]
+    #[cfg(all(feature = "tee", not(feature = "cca")))]
     let qboot_bundle = vm_resources
         .qboot_bundle()
         .ok_or(StartMicrovmError::MissingKernelConfig)?;
```
Can't be `cca` just a different feature?

MatiasVara commented on src/libkrun/src/lib.rs:
```diff
@@ -1242,6 +1255,61 @@ pub extern "C" fn krun_start_enter(ctx_id: u32) -> i32 {
+                        attributes: m.attributes as u64,
+                        flags: 0,
+                    });
+
```
The problem here is that I still need the index to get the fd.

MatiasVara replied:
I open https://github.com/rust-vmm/vm-memory/issues/315 because I think we need an structure to keep track of the mapping `guestdf_mem` and `regions`.

jakecorrenti replied:
Great idea. I'll keep track of that

MatiasVara commented:
After reviewing the Jake comments I came to the conclusion that this PR should be split otherwise is hard to review:

0. Add `Use create_guest_memfd() and set_user_memory_region2()` only when `tee` is enabled.
1. Add the handling of `KVM_MEMORY_FAULT` only for `tee`, which includes the extra thread for handling shared to private and viceverse.
2. `Tee` should be replaced with `tee and x86-64` since all `tee` is for `x86-64`.
3. Add CCA support. This shall be added with `cca and tee` enabled.

jakecorrenti commented:
> After reviewing the Jake comments I came to the conclusion that this PR should be split otherwise is hard to review:
> 
>     0. Add `Use create_guest_memfd() and set_user_memory_region2()` only when `tee` is enabled.
> 
>     1. Add the handling of `KVM_MEMORY_FAULT` only for `tee`, which includes the extra thread for handling shared to private and viceverse.
> 
>     2. `Tee` should be replaced with `tee and x86-64` since all `tee` is for `x86-64`.

I don't know if this is idiomatic by any means, but could we have something like `tee-x86` and `tee`? `tee-x86` for the x86 specific functionality, and then `tee` for the shared bits? Would that be cleaner than just writing `#[cfg(all(feature = "tee", target_arch = "x86"))]`? Again, I don't know if that's "proper" Rust. 
> 
>     3. Add CCA support. This shall be added with `cca and tee` enabled.
> 
>     4. Eventually some code will be shared for `tee` in `x86-64` and `aarch64` but I can't see it yet.
> 
> 
> @jakecorrenti @tylerfanelli @slp WDYT?

Other than my question above, that all sounds good to me.

tylerfanelli commented:
> After reviewing the Jake comments I came to the conclusion that this PR should be split otherwise is hard to review:
> 
> 0. Add `Use create_guest_memfd() and set_user_memory_region2()` only when `tee` is enabled.

I don't think `set_user_memory_region2` should be `tee` only, as I believe that on newer kernels, `set_user_memory_region` is deprecated.

> 1. Add the handling of `KVM_MEMORY_FAULT` only for `tee`, which includes the extra thread for handling shared to private and viceverse.

I'm addressing that in my SEV-SNP PR currently.

> 2. `Tee` should be replaced with `tee and x86-64` since all `tee` is for `x86-64`.

Makes sense to me.

> 3. Add CCA support. This shall be added with `cca and tee` enabled.

Do you mean `aarch64 and tee`?

MatiasVara commented:
> > After reviewing the Jake comments I came to the conclusion that this PR should be split otherwise is hard to review:
> > ```
> > 0. Add `Use create_guest_memfd() and set_user_memory_region2()` only when `tee` is enabled.
> > 
> > 1. Add the handling of `KVM_MEMORY_FAULT` only for `tee`, which includes the extra thread for handling shared to private and viceverse.
> > 
> > 2. `Tee` should be replaced with `tee and x86-64` since all `tee` is for `x86-64`.
> > ```
> 
> I don't know if this is idiomatic by any means, and it depends on whether you can have target arch-specific features, but could we have something like `tee-x86` and `tee`? `tee-x86` for the x86 specific functionality, and then `tee` for the shared bits? Would that be cleaner than just writing `#[cfg(all(feature = "tee", target_arch = "x86_64"))]`? Again, I don't know if that's "proper" Rust or even possible.
> 
I do not know either. I though that `tee and x86` would have code that `tdx` and `snv` share, if that exists.

> > ```
> > 3. Add CCA support. This shall be added with `cca and tee` enabled.
> > 
> > 4. Eventually some code will be shared for `tee` in `x86-64` and `aarch64` but I can't see it yet.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > @jakecorrenti @tylerfanelli @slp WDYT?
> 
> Other than my question above, that all sounds good to me.

MatiasVara commented:
> > After reviewing the Jake comments I came to the conclusion that this PR should be split otherwise is hard to review:
> > 
> > 0. Add `Use create_guest_memfd() and set_user_memory_region2()` only when `tee` is enabled.
> 
> I don't think `set_user_memory_region2` should be `tee` only, as I believe that on newer kernels, `set_user_memory_region` is deprecated.
> 

That's OK. I could just leave `set_user_memory_region2` for the future.

> > 1. Add the handling of `KVM_MEMORY_FAULT` only for `tee`, which includes the extra thread for handling shared to private and viceverse.
> 
> I'm addressing that in my SEV-SNP PR currently.
> 

OK but I need to split those chunks from my PR anyway.

> > 2. `Tee` should be replaced with `tee and x86-64` since all `tee` is for `x86-64`.
> 
> Makes sense to me.
> 
> > 3. Add CCA support. This shall be added with `cca and tee` enabled.
> 
> Do you mean `aarch64 and tee`?

Yes for the moment.

MatiasVara committed:
Use kvm-ioctl and kvm-bindings for cca

This commit will be removed when changes are in upstream.

Signed-off-by: Matias Ezequiel Vara Larsen <mvaralar@redhat.com>

MatiasVara committed:
Enable virtio-fs in cca

This commit will be remove when cca use virtio-blk. For the moment,
enable virtio-fs in cca.

Signed-off-by: Matias Ezequiel Vara Larsen <mvaralar@redhat.com>

MatiasVara commented:
> After reviewing the Jake comments I came to the conclusion that this PR should be split otherwise is hard to review:
> 
> 0. Add `Use create_guest_memfd() and set_user_memory_region2()` only when `tee` is enabled.
> 1. Add the handling of `KVM_MEMORY_FAULT` only for `tee`, which includes the extra thread for handling shared to private and viceverse.
> 2. `Tee` should be replaced with `tee and x86-64` since all `tee` is for `x86-64`.
> 3. Add CCA support. This shall be added with `cca and tee` enabled.
> 4. Eventually some code will be shared for `tee` in `x86-64` and `aarch64` but I can't see it yet.
> 
> @jakecorrenti @tylerfanelli @slp WDYT?

I just split the PR into a few commits. It is the same code though. The first three commits are independent of cca and should be shared by all tees.

MatiasVara commented:
I've Just rebased the PR and tried it on v8 series.

MatiasVara commented:
I just rebased and tested on v8. Also, I am now relying on the rust-vmm/kvm repo.
